python/lsst/daf/butler/cli/utils.py

```diff
@@ -65,3 +130,82 @@ def to_upper(context, param, value):
         A copy of the passed-in value, converted to upper case.
     """
     return value.upper()
+
+
+def printFunctionInfo(func, *args, **kwargs):
+    """For unit testing butler subcommand call execution, write a dict to
+    stdout that formats information about a funciton call into a dict that can
+    be evaluated by `verifyFunctionInfo`
+
+    Parameters
+    ----------
+    func : function
+        The function that has been called, whose name should be written.
+    args : [`str`]
+        The values of the arguments the function was called with.
+    kwags : `dict` [`str`, `str`]
+        The names and values of the kwargs the function was called with.
+    """
+    print(dict(function=func.__name__,
+               args=args,
+               kwargs=kwargs))
+
+
+def verifyFunctionInfo(testSuite, output, function, expectedArgs, expectedKwargs):
+    """For unit testing butler subcommand call execution, compare a dict that
+    has been printed to stdout to expected data.
+
+    Parameters
+    ----------
+    testSuite : `unittest.Testsuite`
+        The test suite that is executing a unit test.
+    output : `str`
+        The dict that has been printed to stdout. It should be formatted to
+        re-instantiate by calling `eval`.
+    function : `str`
+        The name of the function that was was expected to have been called.
+    expectedArgs : [`str`]
+        The values of the arguments that should have been passed to the
+        function.
+    expectedKwargs : `dict` [`str`, `str`]
+        The names and values of the kwargs that should have been passsed to the
+        funciton.
+    """
+    calledWith = eval(output)
+    testSuite.assertEqual(calledWith['function'], function)
+    testSuite.assertEqual(calledWith['args'], expectedArgs)
+    testSuite.assertEqual(calledWith['kwargs'], expectedKwargs)
+
+
+def cli_handle_exception(func, *args, **kwargs):
+    """Wrap a function call in an exception handler that raises a
+    ClickException if there is an Exception.
+
+    Also provides support for unit testing by testing for an environment
+    variable, and if it is present prints the function name, args, and kwargs
+    to stdout so they can be read and verified by the unit test code.
+
+    Parameters
+    ----------
+    func : function
+        A function to be called and exceptions handled. Will pass args & kwargs
+        to the function.
+
+    Returns
+    -------
+    The result of calling func.
+
+    Raises
+    ------
+    click.ClickException
+        An exception to be handled by the Click CLI tool.
+    """
+    # "DAF_BUTLER_MOCK" matches the key in the variable DAF_BUTLER_MOCK,
+    # defined in the top of this file.
+    if "DAF_BUTLER_MOCK" in os.environ:
+        printFunctionInfo(func, *args, **kwargs)
+        return
+    try:
```

Comment:
I tried to implement this with a unittest.mock, but couldn't find a way to get the called mock object back out to the calling script for verification. I couldn't pass it in because butler wants to import the command, and even when replacing the object in sys.modules, import still provides a copy of the class, not the class itself :-(